pkg/provider/azure_controller_vmss.go

```diff
@@ -263,47 +263,40 @@ func (ss *ScaleSet) DetachDisk(ctx context.Context, nodeName types.NodeName, dis
 
 // UpdateVM updates a vm
 func (ss *ScaleSet) UpdateVM(ctx context.Context, nodeName types.NodeName) error {
+	future, err := ss.UpdateVMAsync(ctx, nodeName)
+	if err != nil {
+		return err
+	}
+
+	return ss.WaitForUpdateResult(ctx, future, nodeName, "update_vm")
+}
+
+// UpdateVMAsync updates a vm asynchronously
+func (ss *ScaleSet) UpdateVMAsync(ctx context.Context, nodeName types.NodeName) (*azure.Future, error) {
 	vmName := mapNodeNameToVMName(nodeName)
 	vm, err := ss.getVmssVM(vmName, azcache.CacheReadTypeDefault)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	nodeResourceGroup, err := ss.GetNodeResourceGroup(vmName)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	var updateResult *compute.VirtualMachineScaleSetVM
-	var rerr *retry.Error
-
+	// Invalidate the cache right after updating
 	defer func() {
-		// If there is an error with Update operation,
-		// invalidate the cache
-		if rerr != nil {
-			_ = ss.DeleteCacheForNode(vmName)
-			return
-		}
-
-		// Update the cache with the updated result only if its not nil
-		// and contains the VirtualMachineScaleSetVMProperties
-		if updateResult != nil && updateResult.VirtualMachineScaleSetVMProperties != nil {
-			if err := ss.updateCache(vmName, nodeResourceGroup, vm.VMSSName, vm.InstanceID, updateResult); err != nil {
```

Comment:
not expected code change, pls rebase to master branch

Reply:
For async operation, I believe the cache update would happen at the `WaitForUpdateResult`

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/e2197aaae8487b0d64530791d05e423e6081cacb/pkg/provider/azure_controller_vmss.go#L144-L153

Reply:
there is already a PR (with refactor) to address this: https://github.com/kubernetes-sigs/cloud-provider-azure/pull/2924, pls revert this change and only focus on vm update retries, thanks.

Reply:
The removal of the cache update in the defer statement is a part of changes that `UpdateAsync` function brings. We can't keep the cache update in the defer here as there won't be any result to update the cache with in this async call.